test/core/end2end/tests/retry_streaming_after_commit.cc

```diff
@@ -16,295 +16,108 @@
+  Step();
   // Client receives a second message.
-  memset(ops, 0, sizeof(ops));
-  op = ops;
-  op->op = GRPC_OP_RECV_MESSAGE;
-  op->data.recv_message.recv_message = &response2_payload_recv;
-  op++;
-  error = grpc_call_start_batch(c, ops, static_cast<size_t>(op - ops),
-                                grpc_core::CqVerifier::tag(5), nullptr);
-  GPR_ASSERT(GRPC_CALL_OK == error);
-  cqv.Expect(grpc_core::CqVerifier::tag(5), true);
-  cqv.Verify();
-
+  IncomingMessage server_message2;
+  c.NewBatch(5).RecvMessage(server_message2);
+  Expect(5, true);
+  Step();
   // Client receives status.
-  memset(ops, 0, sizeof(ops));
-  op = ops;
-  op->op = GRPC_OP_RECV_STATUS_ON_CLIENT;
-  op->data.recv_status_on_client.trailing_metadata = &trailing_metadata_recv;
-  op->data.recv_status_on_client.status = &status;
-  op->data.recv_status_on_client.status_details = &details;
-  op++;
-  error = grpc_call_start_batch(c, ops, static_cast<size_t>(op - ops),
-                                grpc_core::CqVerifier::tag(1), nullptr);
-  GPR_ASSERT(GRPC_CALL_OK == error);
-  cqv.Expect(grpc_core::CqVerifier::tag(1), true);
-  cqv.Verify();
-
-  GPR_ASSERT(status == GRPC_STATUS_ABORTED);
-  GPR_ASSERT(0 == grpc_slice_str_cmp(details, "xyz"));
-  GPR_ASSERT(0 == grpc_slice_str_cmp(call_details.method, "/service/method"));
-  GPR_ASSERT(was_cancelled == 0);
-
-  grpc_slice_unref(details);
-  grpc_metadata_array_destroy(&initial_metadata_recv);
-  grpc_metadata_array_destroy(&trailing_metadata_recv);
-  grpc_metadata_array_destroy(&request_metadata_recv);
-  grpc_call_details_destroy(&call_details);
-  grpc_byte_buffer_destroy(request_payload);
-  grpc_byte_buffer_destroy(request2_payload);
-  grpc_byte_buffer_destroy(response_payload);
-  grpc_byte_buffer_destroy(response2_payload);
-  GPR_ASSERT(byte_buffer_eq_slice(request_payload_recv, request_payload_slice));
-  grpc_byte_buffer_destroy(request_payload_recv);
-  GPR_ASSERT(
-      byte_buffer_eq_slice(request2_payload_recv, request2_payload_slice));
-  grpc_byte_buffer_destroy(request2_payload_recv);
-  GPR_ASSERT(
-      byte_buffer_eq_slice(response_payload_recv, response_payload_slice));
-  grpc_byte_buffer_destroy(response_payload_recv);
-  GPR_ASSERT(
-      byte_buffer_eq_slice(response2_payload_recv, response2_payload_slice));
-  grpc_byte_buffer_destroy(response2_payload_recv);
-  grpc_call_unref(c);
-  grpc_call_unref(s);
-}
-
-void retry_streaming_after_commit(const CoreTestConfiguration& config) {
-  GPR_ASSERT(config.feature_mask & FEATURE_MASK_SUPPORTS_CLIENT_CHANNEL);
-  test_retry_streaming_after_commit(config);
+  IncomingStatusOnClient server_status;
+  c.NewBatch(1).RecvStatusOnClient(server_status);
+  Expect(1, true);
+  Step();
+  EXPECT_EQ(server_status.status(), GRPC_STATUS_ABORTED);
+  EXPECT_EQ(server_status.message(), "xyz");
+  EXPECT_EQ(s.method(), "/service/method");
+  EXPECT_FALSE(client_close.was_cancelled());
+  EXPECT_EQ(client_message.payload(), "foo");
```

Comment:
These expectations can move up to be right under the respective recv_message op completions.